### PR TITLE
DBZ-3141 Support for signalling table

### PR DESCRIPTION
**Pull request by jpechane**
https://issues.redhat.com/browse/DBZ-3141

Vitess failure handled in https://github.com/debezium/debezium-connector-vitess/pull/27

### CODE REVIEW

**Comment by jpechane** [comment]
Oracle UX can be improved via https://github.com/debezium/debezium-examples/pull/161

**Comment by jpechane** [comment]
Fixes https://issues.redhat.com/browse/DBZ-2919

File: debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java [hunk 1]
```diff
@@ -410,6 +410,13 @@ public static BinaryHandlingMode parse(String value, String defaultValue) {
             .withValidation(Field::isPositiveInteger)
             .withDescription("The maximum number of threads used to perform the snapshot.  Defaults to 1.");
 
+    public static final Field SIGNAL_DATA_COLLECTION = Field.create("signal.data.collection")
+            .withDisplayName("Signaling data collection")
+            .withType(Type.STRING)
+            .withWidth(Width.MEDIUM)
+            .withImportance(Importance.MEDIUM)
+            .withDescription("The name of the data collection that is used to send signlas/commands to Debezium. Signaling is disabled when not set.");
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java:418]:
```suggestion
            .withDescription("The name of the data collection that is used to send signals/commands to Debezium. Signaling is disabled when not set.");
```

File: debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java [hunk 2]
```diff
@@ -431,7 +438,8 @@ public static BinaryHandlingMode parse(String value, String defaultValue) {
                     TOMBSTONES_ON_DELETE,
                     SOURCE_STRUCT_MAKER_VERSION,
                     Heartbeat.HEARTBEAT_INTERVAL,
-                    Heartbeat.HEARTBEAT_TOPICS_PREFIX)
+                    Heartbeat.HEARTBEAT_TOPICS_PREFIX,
+                    SIGNAL_DATA_COLLECTION)
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java:442]: Should really all support it by default? I'd rather have expected only for the Oracle connector to add this?

File: debezium-core/src/main/java/io/debezium/pipeline/signal/Signal.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+            return "Payload [id=" + id + ", type=" + type + ", data=" + data + "]";
+        }
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(Signal.class);
+
+    private final CommonConnectorConfig connectorConfig;
+    private final String signalDataCollectionId;
+
+    private final Map<String, Action> signalActions = new HashMap<>();
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/pipeline/signal/Signal.java:82]: Is this accessed concurrently? Let's add some `@ThreadSafe` etc. annotations.

**Review by gunnarmorling — COMMENTED** [review]
@jpechane, a few comments inline. So no usage of that for Oracle DDL changes yet, only the basic infra, right?

File: debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java (hunk 2 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java:442]: @gunnarmorling Yes, I'd really like to make it available for all by default. Even the default logging action could be useful for lal connecotrs.

**Comment by gunnarmorling** [comment]
> Even the default logging action

What is this exactly?

>

**Comment by jpechane** [comment]
@gunnarmorling https://github.com/debezium/debezium/pull/2182/files#diff-b20c37494fe1f155bb2197d5d428273e45bbd39e90ea937f52de1860f4aa6dbb

**Comment by gunnarmorling** [comment]
@jpechane, ok, so what's the purpose of using the signal table for Postgres?

**Comment by jpechane** [comment]
@gunnarmorling The prupose is twofold

- have a regularly running IT test
- solve the issue with PK definition change

**Comment by gunnarmorling** [comment]
Ah, gotcha. So we amend the schema evolution story for PG, too. Ok, can you then amend the documentation accordingly, too?

**Comment by jpechane** [comment]
@gunnarmorling Could you please merge the PR and I'll post the docs in a separate one so we can get tests passing again?

**Comment by gunnarmorling** [comment]
@jpechane, are you still planning to push a commit with doc updates for this one?

File: debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/SignalsIT.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+import org.junit.Test;
+
+import io.debezium.config.Configuration;
+import io.debezium.connector.oracle.OracleConnectorConfig.SnapshotMode;
+import io.debezium.connector.oracle.util.TestHelper;
+import io.debezium.embedded.AbstractConnectorTest;
+import io.debezium.util.Testing;
+
+/**
+ * Integration test to check transaction metadata.
```
Comment (gunnarmorling) [inline debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/SignalsIT.java:29]: Copy/paste glitch?

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/SignalsIT.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+        waitForAvailableRecords(100, TimeUnit.MILLISECONDS);
+        // there shouldn't be any snapshot records
+        assertNoRecordsToConsume();
+
+        // insert and verify a new record
+        TestHelper.execute(INSERT_STMT);
+
+        // Insert the signal record - add 'aa' column to PK fields
+        TestHelper.execute("INSERT INTO s1.debezium_signal VALUES('1', 'schema-changes', '{\"database\": \"postgres\", \"changes\": [{\n"
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/SignalsIT.java:131]: I wish we'd use Java 13+ and its test blocks for tests...

**Review by gunnarmorling — COMMENTED** [review]
LGTM, great work. I'll apply that one JavaDoc fix and merge. Thanks a lot, @jpechane!

**Review by gunnarmorling — APPROVED** [review]

**Pull request merged**